aosp_diff/celadon_ivi/frameworks/av/services/camera/libcameraservice/0001-Camera-fix-uvc-cameraid-mNumberOfCameras-issue.patch

```diff
@@ -0,0 +1,81 @@
+From 4d0a18565d5fc935fe1e0df771f734cb05d2344b Mon Sep 17 00:00:00 2001
+From: neofang7 <neo.fang@intel.com>
+Date: Wed, 31 May 2023 02:04:12 +0000
+Subject: [PATCH] Camera: fix uvc cameraid>mNumberOfCameras issue
+
+The cameraId of uvc camera is video device number + CameraIdOffset,
+which is larger than the real camera numbers and causes Invalid cameraId errors.
+Hence, Android app failed to use uvc camera through external cameraHAL.
+The fix contains two parts:
+- libcameraservice.so: change the condition check of cameraId, iterate check app
+  required cameraId with deviceIds;
+- Camera Jni: remove cameraId > NumberOfCameras condition. This patch is delivered
+  in base/core/jni project.
+
+Bug: OAM-109947
+Test: verify camera functionalities with meeting/camera apps
+---
+ .../camera/libcameraservice/CameraService.cpp | 28 ++++++++++++-------
+ 1 file changed, 18 insertions(+), 10 deletions(-)
+
+diff --git a/services/camera/libcameraservice/CameraService.cpp b/services/camera/libcameraservice/CameraService.cpp
+index 701bec2313..682b543952 100644
+--- a/services/camera/libcameraservice/CameraService.cpp
++++ b/services/camera/libcameraservice/CameraService.cpp
+@@ -25,6 +25,7 @@
+ #include <cstring>
+ #include <ctime>
+ #include <string>
++#include <iostream>
+ #include <sys/types.h>
+ #include <inttypes.h>
+ #include <pthread.h>
+@@ -660,14 +661,7 @@ Status CameraService::getCameraInfo(int cameraId,
+         return STATUS_ERROR(ERROR_DISCONNECTED,
+                 "Camera subsystem is not available");
+     }
+-    bool hasSystemCameraPermissions =
+-            hasPermissionsForSystemCamera(CameraThreadState::getCallingPid(),
+-                    CameraThreadState::getCallingUid());
+-    int cameraIdBound = mNumberOfCamerasWithoutSystemCamera;
+-    if (hasSystemCameraPermissions) {
+-        cameraIdBound = mNumberOfCameras;
+-    }
+-    if (cameraId < 0 || cameraId >= cameraIdBound) {
++    if (cameraId < 0) {
+         return STATUS_ERROR(ERROR_ILLEGAL_ARGUMENT,
+                 "CameraId is not valid");
+     }
+@@ -694,13 +688,27 @@ std::string CameraService::cameraIdIntToStrLocked(int cameraIdInt) {
+             getpid() == callingPid) {
+         deviceIds = &mNormalDeviceIds;
+     }
+-    if (cameraIdInt < 0 || cameraIdInt >= static_cast<int>(deviceIds->size())) {
```

Comment:
only this change is sufficient right ? to avoid number of camera with camera id not matching ?

Reply:
Yes, I change the logic of cameraIdInt validation. Left part of the patch will check whether cameraId is in deviceIds. It's invalid if not in deviceIds.